README.md
# hiredis-rb

[![Build Status](https://travis-ci.org/redis/hiredis-rb.svg?branch=master)](https://travis-ci.org/redis/hiredis-rb)

Ruby extension that wraps [hiredis](http://github.com/redis/hiredis). Both
the synchronous connection API and a separate protocol reader are supported.
It is primarily intended to speed up parsing multi bulk replies.

## Install

Install with Rubygems:

    gem install hiredis

## Usage

Hiredis can be used as standalone library, or be used together with redis-rb.
The latter adds in support for hiredis in 2.2.

### redis-rb

To use hiredis from redis-rb, it needs to be available in Ruby's load path.
Using Bundler, this comes down to adding the following lines:

``` ruby
gem "hiredis", "~> 0.6.0"
gem "redis", ">= 3.2.0"
```

To use hiredis with redis-rb, you need to require `redis/connection/hiredis`
before creating a new connection. This makes sure that hiredis will be used
instead of the pure Ruby connection code and protocol parser. Doing so in the
Gemfile is done by adding a `:require` option to the line adding the redis-rb
dependency:

``` ruby
gem "redis", ">= 3.2.0", :require => ["redis", "redis/connection/hiredis"]
```

You can use Redis normally, as you would with the pure Ruby version.

### Standalone: Connection

A connection to Redis can be opened by creating an instance of
`Hiredis::Connection` and calling `#connect`:

``` ruby
conn = Hiredis::Connection.new
conn.connect("127.0.0.1", 6379)
```

Commands can be written to Redis by calling `#write` with an array of
arguments. You can call write more than once, resulting in a pipeline of
commands.

``` ruby
conn.write ["SET", "speed", "awesome"]
conn.write ["GET", "speed"]
```

After commands are written, use `#read` to receive the subsequent replies.
Make sure **not** to call `#read` more than you have replies to read, or
the connection will block indefinitely. You _can_ use this feature
to implement a subscriber (for Redis Pub/Sub).

``` ruby
>> conn.read
=> "OK"

>> conn.read
=> "awesome"
```

When the connection was closed by the server, an error of the type
`Hiredis::Connection::EOFError` will be raised. For all I/O related errors,
the Ruby built-in `Errno::XYZ` errors will be raised. All other errors
(such as a protocol error) result in a `RuntimeError`.

You can skip loading everything and just load `Hiredis::Connection` by
requiring `hiredis/connection`.

### Standalone: Thread Safe Connection
If you want to use `Hiredis::Connection` in a multi-threaded environment, replace `Hiredis::Connection` with `Hiredis::ThreadSafeConnection`.  All threads have their own independent pipeline; e.g. sending a write on one thread, and a read on another will just cause the second thread to block. 

``` ruby
conn = Hiredis::ThreadSafeConnection.new
conn.connect("127.0.0.1", 6379)
Thread.new do
  conn.write ["SET", "hello", "world"]
end
```

Connections can be created lazily, but this can be prevented by persisting your thread pool and passing the optional argument `standby_pool_size` to ``Hiredis::ThreadSafeConnection``.  The example below would create 16 connections on initializations and then pass those connections to threads when they first request a write or read.  Note that if you do not persist your threads, this pool will quickly run out and ``Hiredis::ThreadSafeConnection`` will be forced to dynamically connect.

``` ruby
conn = Hiredis::ThreadSafeConnection standby_pool_size: 16
conn.connect("127.0.0.1", 6379)
```


### Standalone: Reply parser

Only using hiredis for the reply parser can be very useful in scenarios
where the I/O is already handled by another component (such as EventMachine).

You can skip loading everything and just load `Hiredis::Reader` by requiring
`hiredis/reader`.

Use `#feed` on an instance of `Hiredis::Reader` to feed the stream parser with
new data. Use `#read` to get the parsed replies one by one:

``` ruby
>> reader = Hiredis::Reader.new
>> reader.feed("*2\r\n$7\r\nawesome\r\n$5\r\narray\r\n")
>> reader.gets
=> ["awesome", "array"]
```

## Benchmarks

These numbers were generated by running `benchmark/throughput.rb` against
`ruby 1.9.2p180 (2011-02-18 revision 30909) [x86_64-darwin10.6.0]`. The
benchmark compares redis-rb with the Ruby connection and protocol code against
redis-rb with hiredis to handle I/O and reply parsing.

For simple line or bulk replies, the throughput improvement is insignificant,
while the larger multi bulk responses will have a noticeable higher throughput.

                                                            user     system      total        real
    redis-rb:  1x SET pipeline, 10000 times             0.260000   0.140000   0.400000 (  0.726216)
     hiredis:  1x SET pipeline, 10000 times             0.170000   0.130000   0.300000 (  0.602332)
    redis-rb: 10x SET pipeline, 10000 times             1.550000   0.660000   2.210000 (  2.231505)
     hiredis: 10x SET pipeline, 10000 times             0.400000   0.140000   0.540000 (  0.995266)

    redis-rb:  1x GET pipeline, 10000 times             0.270000   0.150000   0.420000 (  0.730322)
     hiredis:  1x GET pipeline, 10000 times             0.170000   0.130000   0.300000 (  0.604060)
    redis-rb: 10x GET pipeline, 10000 times             1.480000   0.640000   2.120000 (  2.122215)
     hiredis: 10x GET pipeline, 10000 times             0.380000   0.130000   0.510000 (  0.925731)

    redis-rb:  1x LPUSH pipeline, 10000 times           0.260000   0.150000   0.410000 (  0.725292)
     hiredis:  1x LPUSH pipeline, 10000 times           0.160000   0.130000   0.290000 (  0.592466)
    redis-rb: 10x LPUSH pipeline, 10000 times           1.540000   0.660000   2.200000 (  2.202709)
     hiredis: 10x LPUSH pipeline, 10000 times           0.360000   0.130000   0.490000 (  0.940361)

    redis-rb:  1x LRANGE(100) pipeline, 1000 times      0.240000   0.020000   0.260000 (  0.307504)
     hiredis:  1x LRANGE(100) pipeline, 1000 times      0.050000   0.020000   0.070000 (  0.100293)

    redis-rb:  1x LRANGE(1000) pipeline, 1000 times     2.100000   0.030000   2.130000 (  2.353551)
     hiredis:  1x LRANGE(1000) pipeline, 1000 times     0.320000   0.030000   0.350000 (  0.472789)

## License

This code is released under the BSD license, after the license of hiredis.


CHANGELOG.md
### 0.6.3 (2018-11-08)

* Fix the packaged gem

### 0.6.2 (2018-11-08)

* Update to use new hiredis 0.14
* Fix bugs with values being garbage collected while still in use

### 0.6.1 (2015-12-29)

* Disable compilation of C extension on unsupported platforms (such as Windows)

### 0.6.0 (2015-02-08)

* Update to use new hiredis 0.12
* Small fixes to make tests work correctly
* Do not test on 1.8.7, 1.9.2 or ree anymore

### 0.5.3

* Add license to gemspec (see #28).

### 0.5.2

* Fix build issue on FreeBSD (see #24).

### 0.5.1

* Fix memory leak for MRI >= 1.9.2 introduced in 0.5.0 (see #22).

### 0.5.0

* Update calls to deprecated Ruby functions with their non-deprecated
  equivalents (see #20 and f85e8c65).

* Update hiredis to 0.11.0.

* Reduced number of objects to garbage collect on Rubinius (see #13).

* Configurable `make` command (see #5).

### 0.4.5

* The protocol reader now forces all strings to be encoded using
  `Encoding.default_external`.

### 0.4.4

* Make tests explicitly require files from the local tree to prevent files from
  the search path to be accidentally required.

### 0.4.3

* Fix bug that caused EAGAIN to be raised after the cumulative time spent
  waiting for the socket to become readable/writable exceeded the
  connection-wide timeout.

### 0.4.2 (unreleased)

* Use patched version of hiredis to support multi bulk depth of 2.

### 0.4.1

* Block indefinitely when timeout is set to zero.

### 0.4.0

* Refactor both the pure Ruby and the native connection class to use
  non-blocking I/O. The code now uses `IO.select` for the pure Ruby connection
  class, and `rb_thread_select` for the native connection class, to detect if a
  socket is readable/writable. This makes the code more portable (w.r.t.
  timeouts on connect/read/write), and more friendly towards threads running in
  the same interpreter (they can now be properly scheduled while hiredis blocks
  on select(2)).

* Add `#flush` method to connection class that flushes the write buffer to the
  socket. This buffer was previously only flushed whenever `#read` was called.

### 0.3.2

* Always statically link to the bundled hiredis version instead of searching
  the system-wide paths.

* Update hiredis to 0.10.0.

### 0.3.1

* Fix bug where one or more arguments passed to #write were garbage collected
  before being appended to the write buffer.

### 0.3.0

* Modify `Connection#connect` and `Connection#connect_unix` to accept an extra
  timeout argument. When connecting times out, `Errno::ETIMEDOUT` is raised.
  The timeout value should be given as number of microseconds to wait.

* Add support for connecting to Unix sockets via `Connection#connect_unix`.

* Drop dependency on redis-rb so it can be used independently, or in another
  library that doesn't require redis-rb.

* Add pure Ruby protocol parser and connection class to use as fallback when
  the extension cannot be loaded. These classes have the same API as the
  extension and use the same unit tests to ensure compatibility.



benchmark/throughput.rb
# Compare performance of redis-rb with and without hiredis
#
# Run with
#
#   $ ruby -Ilib benchmark/throughput.rb
#

require "rubygems"
require "benchmark"
require "redis/connection/hiredis"
require "redis/connection/ruby"
require "redis"

DB = 9

$ruby = Redis.new(:db => DB)
$ruby.client.instance_variable_set(:@connection,Redis::Connection::Ruby.new)
$hiredis = Redis.new(:db => DB)
$hiredis.client.instance_variable_set(:@connection,Redis::Connection::Hiredis.new)

# make sure both are connected
$ruby.ping
$hiredis.ping

# test if db is empty
if $ruby.dbsize > 0
  STDERR.puts "Database \##{DB} is not empty!"
  exit 1
end

def without_gc
  GC.start
  GC.disable
  yield
ensure
  GC.enable
end

def pipeline(b,num,size,title,cmd)
  commands = size.times.map { cmd }

  x = without_gc {
    b.report("redis-rb: %2dx #{title} pipeline, #{num} times" % size) {
      num.times {
        $ruby.client.call_pipelined(commands)
      }
    }
  }

  y = without_gc {
    b.report(" hiredis: %2dx #{title} pipeline, #{num} times" % size) {
      num.times {
        $hiredis.client.call_pipelined(commands)
      }
    }
  }

  puts "%.1fx" % [1 / (y.real / x.real)]
end

Benchmark.bm(50) do |b|
  pipeline(b,10000, 1, "SET", %w(set foo bar))
  pipeline(b,10000,10, "SET", %w(set foo bar))
  puts

  pipeline(b,10000, 1, "GET", %w(get foo))
  pipeline(b,10000,10, "GET", %w(get foo))
  puts

  pipeline(b,10000, 1, "LPUSH", %w(lpush list fooz))
  pipeline(b,10000,10, "LPUSH", %w(lpush list fooz))
  puts

  pipeline(b,1000, 1, "LRANGE(100)", %w(lrange list 0  99))
  puts

  pipeline(b,1000, 1, "LRANGE(1000)", %w(lrange list 0 999))
  puts

  # Clean up...
  $ruby.flushdb
end
